crates/client/sync/src/fetch/mod.rs

```diff
@@ -1,96 +1,111 @@
-        let fetch_stream = (first_block..).take(n_blocks_to_sync.unwrap_or(u64::MAX) as _).map(|block_n| {
-            let provider = Arc::clone(&provider);
-            let cancellation_token = cancellation_token.clone();
-            async move {
-                (
-                    block_n,
-                    fetch_block_and_updates(&backend.chain_config().chain_id, block_n, &provider, &cancellation_token)
-                        .await,
-                )
-            }
-        });
+    if warp_update {
+        let client = jsonrpsee::http_client::HttpClientBuilder::default()
+            .build(format!("http://localhost:{warp_update_port_rpc}"))
+            .expect("Building client");
 
-        // Have 10 fetches in parallel at once, using futures Buffered
-        let mut fetch_stream = stream::iter(fetch_stream).buffered(10);
-        while let Some((block_n, val)) =
-            channel_wait_or_graceful_shutdown(fetch_stream.next(), &cancellation_token).await
-        {
-            match val {
-                Err(FetchError::Sequencer(SequencerError::StarknetError(StarknetError {
-                    code: StarknetErrorCode::BlockNotFound,
-                    ..
-                }))) => {
-                    tracing::info!("🥳 The sync process has caught up with the tip of the chain");
-                    break;
-                }
-                val => {
-                    if fetch_stream_sender.send(val?).await.is_err() {
-                        // join error
-                        break;
-                    }
-                }
-            }
+        if client.ping().await.is_err() {
+            tracing::error!("❗ Failed to connect to warp update sender on http://localhost:{warp_update_port_rpc}");
+            ctx.cancel_global();
+            return Ok(());
+        }
+
+        let provider = Arc::new(GatewayProvider::new(
+            Url::parse(&format!("http://localhost:{warp_update_port_fgw}/gateway/"))
+                .expect("Failed to parse warp update sender gateway url. This should not fail in prod"),
+            Url::parse(&format!("http://localhost:{warp_update_port_fgw}/feeder_gateway/"))
+                .expect("Failed to parse warp update sender feeder gateway url. This should not fail in prod"),
+        ));
+
+        let save = config.sync_parallelism;
+        let available_parallelism = std::thread::available_parallelism()
+            .unwrap_or(NonZeroUsize::new(1usize).expect("1 should always be in usize bound"));
+        config.sync_parallelism = Into::<usize>::into(available_parallelism) * 2;
+
+        let next_block = match sync_blocks(backend.as_ref(), &provider, &ctx, &config).await? {
+            SyncStatus::Full(next_block) => next_block,
+            SyncStatus::UpTo(next_block) => next_block,
+        };
+
+        if client.shutdown().await.is_err() {
+            tracing::error!("❗ Failed to shutdown warp update sender");
```

Comment:
isn't this an unrecoverable error?

Reply:
Yep, will make it one :+1

Reply:
This has been updated to behave the same as `if client.ping().await.is_err()`.